hal/src/common/thumbv6m/sercom/v2/spi.rs

```diff
@@ -0,0 +1,1768 @@
+impl_dipo_dopo!((NoneT, Pad3, Pad1, Pad2): (0, 2));
+impl_dipo_dopo!((NoneT, Pad3, Pad1, NoneT): (0, 2));
+
+//impl_dipo_dopo!((Pad0, NoneT, Pad3, Pad1): (0, 3));
+//impl_dipo_dopo!((Pad0, NoneT, Pad3, NoneT): (0, 3));
+impl_dipo_dopo!((NoneT, Pad0, Pad3, Pad1): (0, 3));
+impl_dipo_dopo!((NoneT, Pad0, Pad3, NoneT): (0, 3));
+
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, Pad2): (1, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, NoneT): (1, 0));
+
+impl_dipo_dopo!((Pad1, Pad2, Pad3, NoneT): (1, 1));
+impl_dipo_dopo!((Pad1, NoneT, Pad3, NoneT): (1, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, Pad1): (1, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, NoneT): (1, 1));
+
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, Pad2): (1, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, NoneT): (1, 2));
+
+impl_dipo_dopo!((Pad1, Pad0, Pad3, NoneT): (1, 3));
+//impl_dipo_dopo!((Pad1, NoneT, Pad3, NoneT): (1, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, Pad1): (1, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, NoneT): (1, 3));
+
+impl_dipo_dopo!((Pad2, Pad0, Pad1, NoneT): (2, 0));
+impl_dipo_dopo!((Pad2, NoneT, Pad1, NoneT): (2, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, Pad2): (2, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, NoneT): (2, 0));
+
+impl_dipo_dopo!((Pad2, NoneT, Pad3, Pad1): (2, 1));
+impl_dipo_dopo!((Pad2, NoneT, Pad3, NoneT): (2, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, Pad1): (2, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, NoneT): (2, 1));
+
+impl_dipo_dopo!((Pad2, Pad3, Pad1, NoneT): (2, 2));
+//impl_dipo_dopo!((Pad2, NoneT, Pad1, NoneT): (2, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, Pad2): (2, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, NoneT): (2, 2));
+
+impl_dipo_dopo!((Pad2, Pad0, Pad3, Pad1): (2, 3));
+impl_dipo_dopo!((Pad2, Pad0, Pad3, NoneT): (2, 3));
+//impl_dipo_dopo!((Pad2, NoneT, Pad3, Pad1): (2, 3));
+//impl_dipo_dopo!((Pad2, NoneT, Pad3, NoneT): (2, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, Pad1): (2, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, NoneT): (2, 3));
+
+impl_dipo_dopo!((Pad3, Pad0, Pad1, Pad2): (3, 0));
+impl_dipo_dopo!((Pad3, NoneT, Pad1, Pad2): (3, 0));
+impl_dipo_dopo!((Pad3, Pad0, Pad1, NoneT): (3, 0));
+impl_dipo_dopo!((Pad3, NoneT, Pad1, NoneT): (3, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, Pad2): (3, 0));
+//impl_dipo_dopo!((NoneT, Pad0, Pad1, NoneT): (3, 0));
+
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, Pad1): (3, 1));
+//impl_dipo_dopo!((NoneT, Pad2, Pad3, NoneT): (3, 1));
+
+//impl_dipo_dopo!((Pad3, NoneT, Pad1, Pad2): (3, 2));
+//impl_dipo_dopo!((Pad3, NoneT, Pad1, NoneT): (3, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, Pad2): (3, 2));
+//impl_dipo_dopo!((NoneT, Pad3, Pad1, NoneT): (3, 2));
+
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, Pad1): (3, 3));
+//impl_dipo_dopo!((NoneT, Pad0, Pad3, NoneT): (3, 3));
+
+//=============================================================================
+// Pads
+//=============================================================================
+
+/// Encapsulate the set of pads for an SPI peripheral
```

Comment:
Encapsulates

Reply:
I meant it in the imperative mood. But I may not have been consistent across  the docs. I'll check it and make sure

Reply:
Until you prompted me here, I hadn't actually thought that much about how I write documentation. It looks like I generally stick to two approaches. Either I lead with a descriptive noun phrase and then describe what it does, e.g.

> Type-level `enum` representing the SPI transaction length, in bytes

Or I lead with an imperative that describes what it does, e.g.

> Control the `DIPO` field as a function of the [`PadNum`] type

I think I use the latter when I don't have any meaningful additions to the noun. For example,

> Trait that controls the `DIPO` field as a function of the [`PadNum`] type

is worse, because "Trait that" adds no real value. We already know it's a trait. Instead, I opt for the imperative, because it's usually terse. I think the same reasoning applies for using the imperative in Git commit messages, which I follow.

Reply:
I dont feel strongly about this :joy: I mentioned it as I'm used to the Go style guide, which suggests all type-level doc strings should be full sentences (ie: subject + verb + what) even if it reads verbose.

Your logic makes sense

Reply:
Interesting, I'll have to read up on it. It probably sounds crazy, but I really like to have some justification for every decision, no matter how ultimately arbitrary it is. Ultimate bikeshedding I guess :p 

I always like to read arguments for/against certain styles and that sort of thing. I find it interesting how people reason about stuff.